### DISCUSSION TITLE
Ability to pause/rewind live data sources (DVR)

### DISCUSSION THREAD

**Discussion by amacneil** [foxglove/community#220] [Q&A]
Currently, studio supports connecting to either a live data source, or loading recorded data from a file or remote server. 

We have heard from some users that they would like the ability to pause/rewind playback of live data sources, similar to [DVR](https://en.wikipedia.org/wiki/Digital_video_recorder)/YouTube Live functionality.

The most simple possible implementation of this would use an in-memory cache (obviously limiting how much can be stored), and only cache topics that are currently visible/subscribed.

Follow up work could include:

- Supporting a disk-cache (IndexedDB or local filesystem) for longer caching
- Subscribing to non-visible topics (UI could be challenging)
- Ability to save the recorded file to disk (maybe this should be a totally separate feature? discussed in https://github.com/foxglove/community/issues/107 but we have not had a lot of interest)

**Comment by RMichaelSwan** [comment]
+1 on this. My team of ~15 engineers would much more easily make the switch to Foxglove right now if this was implemented. We have an internal UI which does this caching, though I am not allowed to share the code presently. It supports using local disk caching or in-memory caching of rosbridge data based on whether it is running as a node app.
